Revise 0315 - Add RPC conflict management (#1086)
smartdevicelink/sdl_evolution: proposals/0315-Add-RPC-Conflict-Management.md

* Proposal: [SDL-0315](0315-Add-RPC-Conflict-Management.md)
* Author: [Akihiro Miyazaki (Nexty)](https://github.com/Akihiro-Miyazaki)
* Status: **Returned for Revisions**
* Impacted Platforms: [ Core / HMI ]

## Introduction

> This proposal is to add a function to SDL Core that manages concurrent RPC conflicts according to priority. This will allow OEMs to receive the expected request from SDL Core during RPC conflict by modifying the configuration file in which the priority is set.

Reviewer: Reference to the "configuration file" should be removed, and Introduction should be further updated to reflect revised proposed solution.

Author: @theresalech -san,
Thank you for your review again.
I will modify to change from "configuration file" to " configuration table in the policy table".
Does  this way matched your advise?
Also, I will modify to add the "/ Policy Server" onto Impacted Platforms.
Anyway, I will modify this proposal and make a commit it.
Could you please confirm it again?
Thank you.

Reviewer: I've provided a suggested change to the `Introduction` section.

To solve these problems, we propose to add a new RPC conflict management function to SDL Core.

## Proposed solution
To implement the RPC conflict management, we will add a new RPC conflict management module, `InterruptManager` to SDL Core, and add `InterruptManagerConfig` to policy table.

> RPC priority table and `App Priority` table are specified in `InterruptManagerConfig`. RPC priority table and `App Priority` table are tables that each set the priority of RPC and app. By modifying `InterruptManagerConfig`, an OEM can receive the expected request from SDL Core during RPC conflict. On the other hand, `InterruptManager` reads `InterruptManagerConfig` in policy table during the SDL Core startup and builds the two tables mentioned above based on their settings. When an RPC conflict occurs, the `InterruptManager` first determines the RPC with a high priority according to the RPC priority table. However, if two competing RPCs have the same priority, then the RPC with the higher priority is determined according to the `App Priority` priority table. Therefore, only one modal RPC to be displayed at a time.

Reviewer: Can you please advise which [agreed upon revision](https://github.com/smartdevicelink/sdl_evolution/issues/1068#issuecomment-704467722) requested changing `AppHMIType` priority table to `App Priority` table?

Author: @theresalech -san,
The RPC priority of previous proposal is below.
1. RPC priority
2. appHMIType priority
However, In review, the RPC priority order was chaged as below.
1. Apps with app priority EMERGENCY
2. RPC priority
3. App priority(non-emergency)
4. HMI Status
Since the priority is replaced from appHMIType to app priority, the appHMIType was not used in the priority order. Therefore, I removed the appHMIType section.

#### RPC Conflict Management Configuration Table
The following tables described in InterruptManagerConfig in the Policy table are explained below.

- RPC priority table
- `AppPriority` table
- HMI status table

> The RPC priority is overridden as the order below. Priority order will shift from 1 to 2, from 2 to 3 and from 3 to 4, if the RPCs are same priority.

Reviewer: ```suggestion
The RPC priority will be determined by the order below. Priority order will shift from 1 to 2, from 2 to 3 and from 3 to 4, if the RPCs are same priority.
```

For RPCs with the same priority, the HMI Status table, which is described later, will be used to determine the priority.

<b>HMI Status Table</b><br>
HMI Status table describes the priority for each app level. When a conflict between RPCs with the same priority in the RPC priority table and `App priority` table occurs, the RPC with the higher priority is determined according to HMI Status table. Below are the default settings of HMI Status table.

<b>Table 4.</b> Default settings of HMI Status table

| HMI Level (String) | Priority (INT) | Note (String)    |
|:-:                 |:-:             |:-:               |
| FULL               | 1              | Highest priority |
| LIMITED            | 2              |                  |
| BACKGROUND         | 3              |                  |
| NONE               | 4              | Lowest priority  |

If two RPCs with the same priority are sent, the second RPC sent will be prioritized.

#### RPC Conflict Management Module
The following explains functions of `InterruptManager`.

<b>1.Loading of configuration table</b><br>

> `InterruptManagerConfig` is loaded from policy table during the startup of SDL Core.

Reviewer: Should `InterruptManagerConfig` be `RpcPriority` here?

Author: @theresalech -san,
Thank you for your response.
In this matter, I think OK as it is.
Because I described "RPC priority table and App Priority table are specified in InterruptManagerConfig." in this proposal.
